Guide/routing.markdown

```diff
@@ -0,0 +1,190 @@
+GET /EditPost?postId={postId}&userId={userId} => EditPostAction { postId, userId }
+```
+
+### Parameter Types
+
+AutoRoute by default only works with UUID arguments (and Id types like `Id Post`). If you have a controller like `data HelloWorldController = HelloAction { name :: Text }` where you have text parameter, you have to configure AutoRoute like this:
+
+```haskell
+instance AutoRoute HelloWorldController where
+    parseArgument = parseTextArgument
+```
+
+This way the `name` argument is passed as `Text` instead of `UUID`.
+
+Right now AutoRoute supports only a single type for all given parameters. E.g. an action which takes an UUID and a Text is not supported with AutoRoute right now:
+```haskell
+data HelloController = HelloAction { userId :: !(Id User), name :: Text }
+instance AutRoute HelloController -- This will fail at runtime
+```
+
+This is a technical problem we hope to fix in the future. Until then consider using `param` for the `Text` parameter.
+
+### Request Methods
+
+When a an action is named a certain way, AutoRoute will pick a certain request method for the route. E.g. for a `DeletePostAction` it will only allow requests with the request method `DELETE` because the action name starts with `Delete`. Here is an overview of all naming patterns and their corresponding request method:
+
+```haskell
+Delete_Action => DELETE
+Update_Action => POST, PATCH
+Create_Action => POST
+otherwise     => GET, POST
+```
+
+If you need more strong rules, consider using the other routing APIs available or overriding the `allowedMethodsForAction` like this:
+
+```haskell
+instance AutoRoute HelloWorldController where
+    allowedMethodsForAction "HelloAction" = [ GET ]
+```
+
+### Application Prefix
+
+When using multiple application in your IHP project, e.g. having an admin backend, AutoRoute will prefix with action urls with the application name. E.g. a controller `HelloWorldController` defined in `Admin/Types.hs` will be automatically be prefixed with `/admin` and generate urls such as `/admin/HelloAction`.
+
+This prefixing has special handling for the `Web` module, so that all controllers in the default `Web` module don't have a prefix.
+
+## Custom Routing
+
+Sometimes you have special needs for your routing. For this case IHP provides a lower-level routing API on which `AutoRoute` is built on.
+
+Let's say we have controller like this:
+
+```haskell
+data PostsController = ShowAllMyPostsAction
+```
+
+We want requests to `/posts` to map to `ShowAllMyPostsAction`. For that we need to add a `CanRoute` instance:
+
+```haskell
+instance CanRoute PostsController where
+    parseRoute' = string "/posts" <* endOfInput >> pure ShowAllMyPostsAction
+```
+
+The `parseRoute'` function is a parser which reads an url and returns an action of type `PostsController`. The router uses [attoparsec](https://hackage.haskell.org/package/attoparsec). Take a look at the attoparsec documentation to get a good understanding on how to parse complex urls.
```

Comment:
This should include a table of the most commonly used patterns in URLs, instead of referring users to a parser combinator library!

Reply:
I've reworded this section a  bit. We're mostly using AutoRoute at digitallyinduced, therefore we don't really have that much example codes or patterns we can document. I hope we can add some more documentation as well as better apis for custom routes in the future.